### PR TITLE
Add `AssetChanged` query filter

### PR DESCRIPTION
**Pull request by tychedelia**
# Objective

Implement a new `AssetChanged` query filter that allows users to query for entities whose related assets may have changed.

- Closes https://github.com/bevyengine/bevy/issues/5069
- Unblocks #16420. Currently, `cold-specialization`, a key rendering optimization for unlocking ancillary benefits of the retained render world, is blocked on being unable detect all scenarios in which an entity's mesh/material changes using events and observers. An `AssetChanged` filter will drastically simplify our implementation and be more robust to future changes.

Originally implemented by @nicopap in #5080.

## Solution

- Adds a new `AssetChanged` query filter that initializes a `AssetChanges<A>` resource that tracks changed assets and ticks in `asset_events`.
- ~Reverts #13343 and changes the api of `get_state` to accept `impl Into<UnsafeWorldCell<'w>>` to allow accessing the `AssetChanges<A>` resource.~
- Adds a `AsAssetId` trait used for newtype handle wrappers (e.g. `Mesh3d`) that allows associating a component with the underlying `Asset` it represents.

## Testing

- Tests are added for `AssetChanged`.
- TBD on performance. We are going to add this `Mesh3d` and `MeshMaterial3d` (etc) in the renderer. Long term wins in render performance this unblocks should swamp tracking overhead for any realistic workload.

## Migration Guide

- The `asset_events` system is no longer public. Users should order their systems relative to the `AssetEvents` system set.

### CODE REVIEW

**Comment by Victoronz** [comment]
Why does this revert the `get_state` PR? In the `get_state` impl of this PR, I only see a call to `resource_id` on `world`, which is already available on `Components`.

**Comment by tychedelia** [comment]
> Why does this revert the `get_state` PR? In the `get_state` impl of this PR, I only see a call to `resource_id` on `world`, which is already available on `&Components`.

Because I didn't know that. Will change it back!

**Comment by Victoronz** [comment]
Separately, how should the resource access in `init_fetch` be reasoned about?
A type being private means it cannot be mentioned in user code, however you do not have to mention a type to obtain it via public API! In this case, this is possible via id-based resource fetching.
`UnsafeWorldCell::get_resource` gives its caller the responsibility to ensure it has access to that resource, however this safety contract cannot be passed on to users of the filter.
For both user, library and other in-engine areas, how should we communicate about this?

File: crates/bevy_asset/src/asset_changed.rs
Comment (pcwalton) [inline crates/bevy_asset/src/asset_changed.rs:78]: What about components that store handles to multiple assets of the same type? We don't need to do that right now, but we might have to revisit it later. In my app I have components with handles to dozens of `AnimationClip`s for example. I think you could maybe just have `AsAssetId`'s method return `-> impl Iterator<Item = AssetId>`.

This isn't blocking, just a suggestion.

File: crates/bevy_asset/src/asset_changed.rs
Comment (pcwalton) [inline crates/bevy_asset/src/asset_changed.rs:106]: Nice, I like the way you document the performance.

**Comment by tychedelia** [comment]
> In this case, this is possible via id-based resource fetching.

How would the user forge an id if they do not have access to the underlying type? `UnsafeWorldCell::get_resource` already documents the aliasing requirements as an unsafe API. If users are iterating or creating random opaque `ComponentId`s and getting mutable access they trivially violate that contract by not knowing what resource they are touching.

**Comment by pcwalton** [comment]
Overall this looks fine to me, but as I mentioned we should have an ECS expert look over the `get_state` signature change.

File: crates/bevy_asset/src/lib.rs
Comment (tychedelia) [reply crates/bevy_asset/src/lib.rs:144]: That's what the other crates seem to do, I'm not unsure about our policy here.

File: crates/bevy_asset/src/asset_changed.rs
Comment (tychedelia) [reply crates/bevy_asset/src/asset_changed.rs:78]: The bigger issue is how to handle the associated type. We might have to use some kind of indirection via UntypedAssetId instead to be able to handle multiple kinds of assets.

File: crates/bevy_asset/src/asset_changed.rs
Comment (chescock) [inline crates/bevy_asset/src/asset_changed.rs:220]: I don't think this line is quite enough.  We need `Query` to report the access in `Access<ArchetypeComponentId>` for it to be used by the scheduler, and it currently does that in `QueryState::update_archetype_component_access()`, which only looks at `component_reads_and_writes()` and `component_writes()`.  

I think you need modify `QueryState::new_with_access()` to look at `resource_reads_and_writes()` and `resource_writes()` and convert them to `ArchetypeComponentId`s.  

I think you also need to modify `FilteredAccess::is_compatible()` to consider resources to conflict even when the filters don't overlap.  (That may need to get reworked in the future for resources-as-components, but the semantics seem reasonable.)  

It would be good to add a unit test that verifies that the access is set up correctly.  Maybe check that `fn system(r: ResMut<AssetChanges<A>>, q: Query<Entity, AssetChanged<A>>)` panics when passed to `assert_system_does_not_conflict`, and again with the parameters reversed.

Comment (tychedelia) [reply crates/bevy_asset/src/asset_changed.rs:220]: Alright thanks for the close review. Will add a test to validate.

File: crates/bevy_asset/src/lib.rs
Comment (alice-i-cecile) [reply crates/bevy_asset/src/lib.rs:144]: Much prefer local allows here.

File: crates/bevy_asset/src/asset_changed.rs
Comment (alice-i-cecile) [inline crates/bevy_asset/src/asset_changed.rs:21]: Basic doc strings for internal types please, especially linking the various parts of the machinery together. It's much easier to review and maintain!

Comment (tychedelia) [reply crates/bevy_asset/src/asset_changed.rs:28]: Added docs also noting the safety implications of this resource.

File: crates/bevy_asset/src/asset_changed.rs
Comment (tychedelia) [reply crates/bevy_asset/src/asset_changed.rs:227]: So just added a test and without making any other changes am getting the following error:
```rust
error[B0001]: Query<MyComponent, AssetChanged<MyComponent>> in system bevy_asset::asset_changed::tests::resource_conflict::system accesses component(s)AssetChanges<MyAsset> in a way that conflicts with a previous system parameter. Consider using `Without<T>` to create disjoint Queries or merging conflicting Queries into a `ParamSet`. See: https://bevyengine.org/learn/errors/b0001
```

Thoughts?

Comment (chescock) [reply crates/bevy_asset/src/asset_changed.rs:227]: Oh, right, `ResMut` does unfiltered access, so `FilteredAccess::is_compatible()` will see that the filters overlap and catch it.  Great!  So that won't be a problem unless someone tries to do *mutable* resource access from a query.  

I think it's just the `Access<ArchetypeComponentId>` that would be an issue, then.  I'm not coming up with a simple way to write a test for that, though.

Comment (tychedelia) [reply crates/bevy_asset/src/asset_changed.rs:227]: So, I think it's just adding the following to `QueryState::update_archetype_component_access`? https://github.com/bevyengine/bevy/pull/16810/commits/d0130a34108ece47c77232d3b4a6e85874090547

File: crates/bevy_asset/src/asset_changed.rs
Comment (alice-i-cecile) [inline crates/bevy_asset/src/asset_changed.rs:1]:
```suggestion
//! Defines the [`AssetChanged`] query filter.
```

File: crates/bevy_asset/src/asset_changed.rs
Comment (alice-i-cecile) [inline crates/bevy_asset/src/asset_changed.rs:3]:
```suggestion
//! Like [`Changed`](bevy_ecs::prelude::Changed), but for [`Asset`]s,
//! and triggers whenever the handle or the underlying asset changes.
```

File: crates/bevy_asset/src/asset_changed.rs
Comment (alice-i-cecile) [inline crates/bevy_asset/src/asset_changed.rs:142]:
```suggestion
/// [`WorldQuery`] state for [`AssetChanged`].
```

File: crates/bevy_asset/src/asset_changed.rs
Comment (alice-i-cecile) [inline crates/bevy_asset/src/asset_changed.rs:126]:
```suggestion
/// [`WorldQuery`] fetch for [`AssetChanged`].
```

File: crates/bevy_ecs/src/query/state.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/state.rs:586]: Oh right, this is an existing soundness bug. Re-adding the tag.

File: crates/bevy_asset/src/asset_changed.rs
Comment (tychedelia) [reply crates/bevy_asset/src/asset_changed.rs:91]: Sorry, I also misunderstood you. Multiple assets of the same type should be much easier to handle. I'd been hung up on if, for example we decided to merge mesh and material as has been discussed before.

File: crates/bevy_asset/src/asset_changed.rs
Comment (chescock) [reply crates/bevy_asset/src/asset_changed.rs:230]: Not quite.  That adds access to components on entities in archetypes matched by the query.  But you want to add access to resources from the world.  Resources on the world have their own `ArchetypeComponentId`s that don't appear in any archetypes.  

If we just want to support this use case of a finite set of read access, I think it's enough to add code to `Query::new_with_access` that does
```rust
for component_id in state.component_access.resource_reads() {
    access.add_resource_read(world.initialize_resource_internal(component_id).id());
}
```

More complete support would need to check for writes and for `reads_all` and `writes_all`, and look a bit like the code in `FilteredResourcesMutParamBuilder` https://github.com/bevyengine/bevy/blob/6178ce93e8537f823685e8c1b617e22ba93696e7/crates/bevy_ecs/src/system/builder.rs#L644

---

The background here is that adding access in a `QueryData` has only ever added access to entities matched by the query.  There has never been a way to declare that you want access to resources from the world.  

But the code you wrote *looks* like it should give access to resources!  And that access is useful!  So I think we should add the plumbing to make it work.  

It might be cleaner to do that in a separate PR, though, and put this one back to what you had originally.  This one would just be unsound until the other one was merged.  If you'd like, I can try to put one together on Monday.

Comment (tychedelia) [reply crates/bevy_asset/src/asset_changed.rs:230]: Thanks @chescock, that would be super helpful. I'm mostly trying to land this to unblock some rendering features that depend on asset change detection and some of the ECS internals are a bit above my paygrade. Appreciate your education here, as it's better building my understanding.

Comment (alice-i-cecile) [reply crates/bevy_asset/src/asset_changed.rs:230]: > But the code you wrote looks like it should give access to resources! And that access is useful! So I think we should add the plumbing to make it work.

Yes please! This is really useful for unlocking indexes as a first class ECS feature.

Comment (chescock) [reply crates/bevy_asset/src/asset_changed.rs:230]: Okay!  #16843

While writing tests there, I also noticed we might need to add
```rust
assert!(!access.access().has_resource_write(component_id), "some useful message");
```
just before the `add_resource_read` call, unless we decide to just ban queries from making mutable resource access.

File: crates/bevy_asset/src/asset_changed.rs
Comment (chescock) [inline crates/bevy_asset/src/asset_changed.rs:172]: Since you already have the `resource_id` in `state`, it would be slightly more efficient to call `get_resource_by_id`.  I think that would look something like: 

```suggestion
        // SAFETY:
        // - `AssetChanges` is private and only accessed mutably in the `AssetEvents` schedule
        // - `resource_id` was obtained from the type ID of `AssetChanges<A::Asset>`.
        let Some(changes) = (unsafe {
            world
                .get_resource_by_id(state.resource_id)
                .map(|ptr| ptr.deref::<AssetChanges<A::Asset>>())
        }) else {
```

File: crates/bevy_asset/src/asset_changed.rs
Comment (tychedelia) [reply crates/bevy_asset/src/asset_changed.rs:241]: I've removed the changes I added and merged your changes from main. I've also added the assertion for write access in the filter.

File: crates/bevy_ecs/src/query/state.rs
Comment (tychedelia) [reply crates/bevy_ecs/src/query/state.rs]: These changes have been removed in favor of @chescock's.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_asset/src/asset_changed.rs
Comment (pcwalton) [inline crates/bevy_asset/src/asset_changed.rs:28]: When you mean "should not be exposed", what do you mean? Changing the fields?

I ask because usually safety invariants are scoped to be module-private, not crate-private.

**Review by pcwalton — APPROVED** [review]
Seems fine, with the caveat that as usual I'm not an expert on the ECS guts.

File: crates/bevy_asset/src/asset_changed.rs
Comment (tychedelia) [reply crates/bevy_asset/src/asset_changed.rs:28]: This mostly is meant to mean don't make this pub because some asks. It's crate scoped mostly due to code organization, having the filter live in its own file is I think preferable to including all of this code in the root lib file where the asset event system lives simply for the sake of keeping it module private.

**Pull request merged**
